### PR TITLE
quantization, bug fix in deconv and graph enet

### PR DESCRIPTION
**Pull request by Wangyida**
## Task list:
- [x] Basic functions and pass the tests on core quantization utilities.
- [x] Quantized convolution layer and pass the tests for q_conv.
- [x] Quantized deconvolution layer and pass the tests for q_deconv.
- [x] Quantized fully connected layer and pass the tests for q_fully_connected.
- [x] Quantized bias inside other kernels.
- [x] Ensure acceptable accuracy on typical examples.
- [x] Add low precision gemm from Google TF as a kernel for matmul.
- [x] Resolve unnecessary quantize-dequantize procedure.
- [x] Add backward propagation for quantization.
 
## Method:
The method is totally the same as described in [Pete's blog for quantization](https://petewarden.com/2016/05/03/how-to-quantize-neural-networks-with-tensorflow/) and my basic codes are modified from [TensorFlow quantization module](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/quantization).

### CODE REVIEW

**Comment by bhack** [comment]
@petewarden We are releasing this under BSD. Can you take a look to this diff to see if there are potential copyright problems.

File: tiny_cnn/config.h [hunk 1]
```diff
@@ -69,6 +69,16 @@ namespace tiny_cnn {
...
+typedef long long int64;
+typedef int int32;
+typedef short int16;
+typedef signed char int8;
+typedef int qint32;
+typedef short qint16;
+typedef signed char qint8;
+typedef unsigned int quint32;
+typedef unsigned short quint16;
+typedef unsigned char  quint8;
```
Comment (patrikhuber) [inline tiny_cnn/config.h]: If you need fixed size types, you should use the ones from `<cstdint>` ([link](http://en.cppreference.com/w/cpp/header/cstdint)), I believe that applies here too. It can otherwise create quite a havoc on certain platforms.

Comment (Wangyida) [inline tiny_cnn/config.h]: Yeah, I got it. I will change them in the next commits, Thanks.

**Comment by edgarriba** [comment]
@Wangyida How do you think that the API should be adapted for quantization: in each layer constructor or maybe as a general parameter in 'session' ?

**Comment by Wangyida** [comment]
@edgarriba I implement the quantization as a stand alone kernel and I think other kernels such as conv, deconv and fc have an option on whether to use this kernel. This option could be defined in the layer files.

**Comment by edgarriba** [comment]
@Wangyida Sure, but are we allowing to user to choose quantized/dequantized version?

**Comment by Wangyida** [comment]
@edgarriba As for TF, there is a standalone quantized_conv2d operation accessable to user. If we follow the style of TF, the quantized operation should be set apart as standalone layers such as quantized_convolutional_layer.h. It will call directly for the quantized kernel rather than an option in the layer parameter.

**Comment by edgarriba** [comment]
@Wangyida it seems okay for a first iteration. We can think later what's the best use of case for the user.

**Comment by Wangyida** [comment]
@edgarriba Of course, working matters much now.

File: tiny_cnn/core/kernels/tiny_quantization_kernel.h [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+template <class T>
+T lowest()  {
+  return std::numeric_limits<T>::is_integer ? (std::numeric_limits<T>::min)() : (-(std::numeric_limits<T>::max)());
+}
+
+// We have to be able to detect and handle overflows in int32, so this function
+// uses doubles and int64's to make sure we have enough room.
+template <class T>
+int64 FloatToQuantizedUnclamped(float input, float range_min, float range_max) {
```
Comment (nyanp) [inline tiny_cnn/core/kernels/tiny_quantization_kernel.h]: @Wangyida 
I prefer ```snake_case``` to ```PascalCase``` for non-member function, just for naming consistency in tiny-cnn. Could  you rename it?

Comment (Wangyida) [inline tiny_cnn/core/kernels/tiny_quantization_kernel.h]: @nyanp Alright, I will change these details together with the quantized convolution layer for a clear review.

Comment (nyanp) [inline tiny_cnn/core/kernels/tiny_quantization_kernel.h]: @Wangyida 
Ok, thanks!

**Comment by Wangyida** [comment]
### A new version: quantization is applicable in convolution now.
@bhack @nyanp  I finished codes for convolutional forward pass quantization, the function style are ```snake_case``` now. I also modified the test files for basic quantization utilities and convolution quantization which is the same as full precision convolution except for the fact that the precision is reduced from ```1E-5``` to ```1E-3```. The algorithm is still similar to TF.
@patrikhuber They are replaced with your advise, Thank you.
@edgarriba I mainly put quantization into another convolutional fernel.
@petewarden Is there any potential copyright problems?

**Comment by bhack** [comment]
@Wangyida Can you put  a task lists in the PR description (github style)? So we have a little bit clear the roadmap on this and what tasks are done.

**Comment by Wangyida** [comment]
@bhack  I updated my first description on this PR, is this suitable?

**Comment by bhack** [comment]
Yes. You can change the markdown with this style https://github.com/blog/1825-task-lists-in-all-markdown-documents

**Comment by Wangyida** [comment]
@bhack Nice, clean format, I have changed them.

**Comment by Wangyida** [comment]
@nyanp Now basic codes for quantized conv and deconv is finished, I have made the test codes for them together with the core quantization utility. I tested them in examples of MNIST training, this could be an initial version. Some improper backends definition of deconvolution is also changed to new backends.

**Comment by Wangyida** [comment]
Accuracy on MNIST examples is nearly 0.5% lower than full precision both in examples with deconvolution layer or without..

**Comment by bhack** [comment]
Excluding batch normalizzation how much [kernels](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/quantization/kernels) do you want to cover?

**Comment by Wangyida** [comment]
@bhack As conv/deconv might be the most time consuming operation and fc might be the most memory consuming operation, I would like to apply quantization to these 3 kernels. Now I just apply quantization on forward pass kernels and the codes need to be optimized to make a real saving of time and memory.

**Comment by Wangyida** [comment]
@bhack I think the operation kernels are correctly implemented, do you have further advises? Here I have 2  imperfections in codes:
Though inputs and weights are quantized as 8 bits unsigned int, operations involving matrix multiplication should are done under 32 bits int because of range expanding.
Outputs pariitcular quantized layer are still dequantized as float for next layer. So the saved model is not based on 8 bit datatype.

**Comment by bhack** [comment]
What do you still want to do in this PR? I think that gemmlowp could be introduced in a new one.

**Comment by Wangyida** [comment]
Yeah, I will add this for fully connected layer and a column vecter based convolutional layer

**Comment by Wangyida** [comment]
@nyanp @bhack Now I added gemmlowp as a new kernel for matmul and I tested this kernel on fully connected layer. I think this PR could be merged now and all 3 tests passed according to quantization. The codes could surely be optimized for efficiency afterwards. I tested most condition for the core algorithms for quantization, some bugs might be fixed when they are found.

**Comment by Wangyida** [comment]
@nyanp In AppVeyor online building, it alarms me that:
```
fatal: no submodule mapping found in .gitmodules for path 'external/gemmlowp'
``` 
How to deal with this? Should I change the location where the gemmlowp functions exist?

**Comment by bhack** [comment]
I think that we need to complete coverage to go on the rigth side of this flow [graph flow](https://petewarden.files.wordpress.com/2016/05/quantization2.png?w=978)

**Comment by patrikhuber** [comment]
@Wangyida I think you forgot `git submodule add`? After you add it, it should pop up in the `.gitmodules` files.

**Comment by Wangyida** [comment]
@bhack OK, I got your idea, this is not implemented yet indeed. I will do this.

**Comment by Wangyida** [comment]
@patrikhuber Thanks a lot!

**Comment by edgarriba** [comment]
@Wangyida Rebase from my last PR (it's only CMake stuff, hope no big conflicts).
Squash commits too in the end.

**Comment by Wangyida** [comment]
@edgarriba OK, I will rebase on this.
Hi all, I just come back from Microsoft Faculty Summit I apologize for my absence for these days. Here I push additional functions for quantization. Now both forward and backward propagation could be used in the network. They can be applied standalone without dependency on each other or used both in the training stage. Tests are provided in all quantized layer and kernels. A single example for quantized application is based on conv and fc layers.

@nyanp I think this could be merged after I rebase from the **Batch Norm** PR.

@bhack For the automatic solver for concatenated quantized layers, I use overloading kernels with different input now. It's still not so fast as tensorflow caused be the code structure of the quantization module in tiny-cnn, but all functions are working correctly and I make the BP process also in quantization mode for learning. The accuracy using all quantized layers are acceptable.

**Comment by bhack** [comment]
We need to cover also batchnorm now that is merged. We could optimize something at structure level? We need to cover concat for enet?

**Comment by Wangyida** [comment]
@bhack Yes, structure should be modified if we want to achieve a better speed. Concat should also be covered for ENet. And for the IOS version of tiny-cnn, I would like to have a try on developing an IOS app or similar application on IOS system, I think that this project could be run on iPhone. Is there any information about Object-C to integrate this project into an iPhone App? I don't know how to transfer the cmake-make-execute process into an app style based on IOS.

**Comment by bhack** [comment]
https://github.com/cristeab/ios-cmake

**Comment by Wangyida** [comment]
this is nice:)

**Comment by Wangyida** [comment]
So I will rebase this directly on mater branch rather than graph branch, is this OK?

**Comment by edgarriba** [comment]
@Wangyida nope, from graph branch.
There are huge modifications in master.

I think it's better let @nyanp rebase master from graph.

**Comment by Wangyida** [comment]
@edgarriba Alright, Thanks.

**Comment by Wangyida** [comment]
Rebase is done, maybe gemm should be avoided in Windows environment because there seems some files used in low precision gemm functions referenced from Google can't be included in Windows compared to linux/Unix.

**Comment by bhack** [comment]
pthread code it is very minimal could be substituted with equivalent c++11 thread. I don't know if upstream will accept this little PR.

**Comment by Wangyida** [comment]
@edgarriba Yes, we can, I tried a prototype of such multi branch net, I will add another example for  this style. Now I have a issue for compiling with cmake generated makefile. When I use ```cmake -DUSE_OPENCV -DBUILD_EXAMPLES .``` to enable examples based on OpenCV visualization, it will be such error:
```
fatal error: 'tiny_cnn/tiny_cnn.h' file not found
``` 
I think this is caused by unproper CMakeLists.txt modification, both mnist_test and deconv_visual couldn't be compiled due to the same error, do you have similar problem?

**Comment by edgarriba** [comment]
@Wangyida this might fix it http://pastebin.com/0kraTigs

**Comment by Wangyida** [comment]
@edgarriba OK, got this, Thanks.

**Comment by Wangyida** [comment]
@nyanp I see a lot of commits merged in graph branch, so should I still rebase on it or rebase on master? I am rebasing on graph.

**Comment by nyanp** [comment]
@Wangyida 
I think we should go on graph branch. There are some conflict between this PR and head of graph branch, and I think it's better to solve this kind of conflict as soon as possible even if we have some bugs in this PR.

> 1.quantization 
> 2.bug fix for deconv 
> 3.codes structure modification
> Should I seperate them into 3 PR?

I'll solve 3.code strcuture problem and accept this PR tomorrow, and I want you to solve SEGV problem as a new PR.

>  I would better add develop serialization functions related to model loader from Tensorflow prepared for loading large model in a new PR on the latest head, what do you think?

It's great idea :) Probably we should decide better initialization scheme / parameter data structure at first (see #236), because it must affect the way of serialization/deserialization.

**Comment by Wangyida** [comment]
@nyanp Nice, I will solve the remaining SEGV in this PR, it will also be OK tomorrow. I think we'd better merge this while it's fine in core functions and develop optimization on quanitzation along with other PR for convenience of suiting new codes structure in time.
As mentioned by @bhack I will try to deal serialization/deserialization with protobuf 3.0 which is friendly in API and used in TensorFlow.

**Comment by bhack** [comment]
We want to put gemmlowp in thirdparty dir? In a git submodule? Or use cmake external?

**Comment by Wangyida** [comment]
For now I put it in ./tiny_cnn/gemmlowp, is this OK?

**Comment by edgarriba** [comment]
@nyanp we should also fix AVX compilation issues and merge all both
branches to master ASAP.

@Wanyida @bhack I would go for a cmake gemmlowp. Besides, why protobuf? We
could use hdf5 for serialization. There's another guy in the gsoc working
on that.

**Comment by bhack** [comment]
We are using prototxt for caffe helper class. We need to use prototxt 3.0 for tensorflow helper class to read net structure and weights. (Here are listed)[https://github.com/BVLC/caffe/pull/2836] some historical motivation of the hdf5 introduction in caffe for serializzation. We need to check if some of that was caused by caffe serializzation design and what was related to protobuf 2.x.

**Comment by nyanp** [comment]
@Wangyida 
Now rebase&merge are done. gemmlowp is moved to ```third_party/gemmlowp```. Could you check the bug in deconv layer? Besides, I found that there are a lot of duplicated code between conv/deconv and quantized/non-quantized. It makes tiny-cnn much harder to maintain, so I want you to cleanup duplicated codes after the bug solved.

**Comment by nyanp** [comment]
@edgarriba 
Sure I'll fix AVX compilation error next. We need to fix all known bugs before merging into master.

**Comment by Wangyida** [comment]
@nyanp OK, I'm solving bugs based on the new codes, seems some former bugfixs in this PR are ignored after rebasing because there are no conflict between commits so I am check thoughout the former commits to find out these remaining bugfixes. There are several math related bugs in the upstream graph branch of deconv. I will try to make the whole codes as clean as possible.

**Comment by nyanp** [comment]
@Wangyida 
I pulled this PR and rebase on my local, solved all conflicts by hand, and squashed it into single commit(6a1f6d7). I've merged your former buxfixes as much as possible, but I might overlook some of them. I'll close this PR. Anyway let's check bugs and fix it in next PR. Thank you for your great works :+1:

**Comment by nyanp** [comment]
@edgarriba 
AVX compilation issues are fixed by 9983245 :+1:

**Comment by Wangyida** [comment]
@nyanp Nice, Thank you, I will have a look on the new head to have a check on 3 of my local SEGV problem. Fast modification!

**Comment by bhack** [comment]
In thirdparty without submodule is there a risk to not be out of sync with [upstream](https://github.com/google/gemmlowp/commits/master)?
